### PR TITLE
SALTO-1125: Fix deserialization of container type annotations

### PR DESCRIPTION
**Pull request by ori-moisis**
This fixes an issue where container types would not get their inner type's annotations
and annotation types when deserialized.

---
_Release Notes_
(salesforce bug fix)
- Fix deploy plan incorrectly showing removals of internalId in CPQ CustomScript fields.

### CODE REVIEW

File: packages/workspace/src/serializer/elements.ts [hunk 1]
```diff
@@ -229,17 +229,36 @@ export const deserialize = async (
...
-    _.keys(element).forEach(k => {
-      _.set(element, k, _.cloneDeepWith(_.get(element, k), v => {
-        if (isType(v)) {
-          return typeMap[v.elemID.getFullName()]
-        }
-        if (isStaticFile(v)) {
-          return staticFiles[v.filepath]
+    // We use cloneDeep for the iteration but we change the objects in-place in order to preserve
+    // references between objects
+    _.cloneDeepWith(element, (value, key, object) => {
```
Comment (ori-moisis) [inline packages/workspace/src/serializer/elements.ts:244]: I tried writing a version of this where I implemented the iteration over the object myself to avoid the "clone" part of "cloneDeepWith" because I assumed we could get some performance gains for it.
Surprisingly some quick profiling showed almost no difference in runtime when loading a fairly large state file.
memory consumption results were inconsistent (which means any gain we may have gotten probably isn't that meaningful).

Given those results I figured it isn't worth the risk of implementing the iteration logic differently and decided to keep using `cloneDeepWith` for the iteration

File: packages/workspace/src/serializer/elements.ts [hunk 2]
```diff
@@ -229,17 +229,36 @@ export const deserialize = async (
...
   elements.forEach(element => {
-    _.keys(element).forEach(k => {
-      _.set(element, k, _.cloneDeepWith(_.get(element, k), v => {
-        if (isType(v)) {
-          return typeMap[v.elemID.getFullName()]
-        }
-        if (isStaticFile(v)) {
-          return staticFiles[v.filepath]
+    // We use cloneDeep for the iteration but we change the objects in-place in order to preserve
+    // references between objects
```
Comment (ori-moisis) [inline packages/workspace/src/serializer/elements.ts:243]: the previous code would only change the immediate properties of the top level element in place and would create a copy of the deeper values.
This would create a situation where we may end up with many copies of things like fields and field annotations.

File: packages/workspace/test/workspace/hidden_values.test.ts [hunk 3]
```diff
@@ -172,7 +172,7 @@ describe('handleHiddenChanges', () => {
     })
   })
 
-  describe('hidden annnotation of field', async () => {
```
Comment (ori-moisis) [inline packages/workspace/test/workspace/hidden_values.test.ts:175]: unrelated fix

File: packages/workspace/test/serializer/serializer.test.ts [hunk 4]
```diff
@@ -30,6 +30,14 @@ describe('State/cache serialization', () => {
...
     elemID: new ElemID('salesforce', 'string'),
     primitive: PrimitiveTypes.STRING,
+    annotationTypes: {
+      anno: BuiltinTypes.STRING,
+      hiddenAnno: BuiltinTypes.HIDDEN_STRING,
+    },
+    annotations: {
+      anno: 'type annotation',
+      hiddenAnno: 'hidden',
+    },
```
Comment (ori-moisis) [inline packages/workspace/test/serializer/serializer.test.ts:40]: there is a test below that serializes and deserializes a list of this type and then compares it to the original value, adding these annotations and annotation types here was enough to reproduce the bug scenario

File: packages/workspace/src/serializer/elements.ts [hunk 5]
```diff
@@ -229,17 +229,38 @@ export const deserialize = async (
   const elementsMap = _.keyBy(elements.filter(isType), e => e.elemID.getFullName())
   const builtinMap = _(BuiltinTypes).values().keyBy(b => b.elemID.getFullName()).value()
   const typeMap = _.merge({}, elementsMap, builtinMap)
+  const resolveType = (type: TypeElement): TypeElement | undefined => {
+    if (isContainerType(type)) {
+      const innerType = resolveType(type.innerType) ?? type.innerType
+      type.setInnerType(innerType)
```
Comment (netama) [inline packages/workspace/src/serializer/elements.ts:235]: just making sure - it seems `setInnerType` also overrides the annotations (and not only the annotation types), is that harmless?

File: packages/workspace/src/serializer/elements.ts [hunk 6]
```diff
@@ -229,17 +229,38 @@ export const deserialize = async (
...
-          return typeMap[v.elemID.getFullName()]
-        }
-        if (isStaticFile(v)) {
-          return staticFiles[v.filepath]
+    // We use cloneDeep for the iteration but we change the objects in-place in order to preserve
+    // references between objects
+    _.cloneDeepWith(element, (value, key, object) => {
+      if (object === undefined || key === undefined) {
+        // We don't get object/key if and only if this is called on the top level element
+        if (isContainerType(value)) {
```
Comment (netama) [inline packages/workspace/src/serializer/elements.ts:247]: nit: is the `if` needed? (it might be, I didn't fully understand from the comments when it's ok not to clone and when not, but I _think_ if the element already exists in `typeMap` we should reuse it?)

File: packages/workspace/src/serializer/elements.ts (hunk 1 above)
Comment (netama) [reply packages/workspace/src/serializer/elements.ts:244]: nit: it might not be the best day to evaluate that (😅), but even with the comment I think it's a little confusing to read with how `cloneDeepWith` is used here - seeing some cases that use `_.set()` and others that `return undefined` that are intended for something different than what customizers are inteded for. wdyt about adding (in lowerdash?) some wrapper function that will be used instead, and can have all the comments and explanations (maybe something like `traverse`, or `transformRawObject` or something if we want it to conform to `transformX`)? it can still call `_.cloneDeepWith` the same way, but I think it might be easier to read this part (but maybe not).

**Review by netama — APPROVED** [review]
LGTM, added a couple of questions, and one readability suggestion that I'm not sure is actually helpful... so feel free to ignore.

File: packages/workspace/src/serializer/elements.ts (hunk 1 above)
Comment (ori-moisis) [reply packages/workspace/src/serializer/elements.ts:244]: when I implemented it I called it `mutateDeep`.
the thing is, if we have such an implementation as a standalone function it really doesn't make sense to me to have it use `cloneDeepWith` (and the [iteration logic](https://github.com/lodash/lodash/blob/2f79053d7bc7c9c9561a30dda202b3dcd2b72b90/.internal/baseClone.js#L157) in `cloneDeepWith` is pretty complicated).
In this case I chose to stick with `cloneDeepWith` to make a minimal change, I don't think it is significantly more confusing then the previous implementation but if you think this is really worse I think I'd rather have a "clean" implementation of `mutateDeep` in lowerdash than having a hacky one that uses `cloneDeepWith` for the iteration.
so, what do you think? should we keep this (maybe if I add a comment above the `_.set` it will be clearer?) or implement a new deep mutation function (and then the implementation here would be very similar to the current implementation)?

File: packages/workspace/src/serializer/elements.ts (hunk 5 above)
Comment (ori-moisis) [reply packages/workspace/src/serializer/elements.ts:235]: not just harmless - it is an important part of the bug fix

File: packages/workspace/src/serializer/elements.ts (hunk 6 above)
Comment (ori-moisis) [reply packages/workspace/src/serializer/elements.ts:247]: so, depends on how you look at it.
in real scenarios we will never have container types in the list of top level elements so this case will never happen outside of tests and you could say this `if` is not needed.
however, because this bug is pretty hard to notice and the test we had did put container types in the element list I figured it was worth supporting that case just in case we ever need to do that.

If you think this is more confusing than useful we can remove this condition (and change the test so it does not put container types in the element list), that was my first instinct as well...

also note that the return value doesn't go anywhere, it is not going to cause us to use something that is not from `typeMap`

File: packages/workspace/src/serializer/elements.ts (hunk 1 above)
Comment (netama) [reply packages/workspace/src/serializer/elements.ts:244]: I'm not sure. I think maybe comments should be enough - also, on 2nd read I think the part that confused me was `return resolveType(value)` ([here](https://github.com/salto-io/salto/pull/1761/files#diff-8d5614671f7c55284e0b34d78d5e49bae18c11acb78ce48c1b9e4ea8348954a2R248)) because the return value is also what we would want clone to return in this case. maybe instead something like 
```
resolveType(value)
return true
```
(or some dummy value other than `true`) would make it clearer that the return value is only used for the traversal? also, maybe having the comment next to the `returns` (explaining why we should/shouldn't continue the recursion) would be clearer than on the `_.set()`s. but again not sure :) so your call

Comment (netama) [reply packages/workspace/src/serializer/elements.ts:244]: actually I guess the function might return `undefined` and we would want to continue the traversal, so we should use the value returned...

File: packages/workspace/src/serializer/elements.ts (hunk 6 above)
Comment (netama) [reply packages/workspace/src/serializer/elements.ts:247]: sorry I didn't mean is the case needed - just that the same `if` is computed inside the function too, so if the function's return value is good enough for the traversal (returns `undefined` in cases when we should continue) then we don't need the check. but I may have missed something

Comment (ori-moisis) [reply packages/workspace/src/serializer/elements.ts:247]: yes, in this case (when `object === undefined`) it means we are looking at a top level element, so if we return `resolveType` and the element is a type it will never return `undefined` and we will never recurse into types.
for container types specifically it is safe not to recurse because `resolveType` does all the relevant replaces.

File: packages/workspace/src/serializer/elements.ts (hunk 1 above)
Comment (ori-moisis) [reply packages/workspace/src/serializer/elements.ts:244]: I could change the return statements to something like `return resolvedType === undefined ? undefined : 'stop recursion'`
a little more complex but might be easier to understand?

File: packages/workspace/src/serializer/elements.ts (hunk 6 above)
Comment (netama) [reply packages/workspace/src/serializer/elements.ts:247]: got it, makes sense. thanks

File: packages/workspace/src/serializer/elements.ts (hunk 1 above)
Comment (netama) [reply packages/workspace/src/serializer/elements.ts:244]: yeah - though I have a feeling someone will refactor this back at some point, so might not be worth the hassle :) either way I think it's not a big issue, so whatever seems the most readable to you

Comment (ori-moisis) [reply packages/workspace/src/serializer/elements.ts:244]: here is the second option:
```typescript
  elements.forEach(element => {
    // We use cloneDeep for the iteration but we change the objects in-place in order to preserve
    // references between objects
    _.cloneDeepWith(element, (value, key, object) => {
      if (object === undefined || key === undefined) {
        // We don't get object/key if and only if this is called on the top level element
        if (isContainerType(value)) {
          resolveType(value)
          return 'stop recursion'
        }
        return undefined
      }
      if (isType(value)) {
        const resolvedType = resolveType(value)
        if (resolvedType !== undefined) {
          _.set(object, key, resolvedType)
          return 'stop recursion'
        }
        return undefined
      }
      if (isStaticFile(value)) {
        _.set(object, key, staticFiles[value.filepath])
        return 'stop recursion'
      }
      return undefined
    })
  })
```

reply with 👍  and I'll commit this option, reply with 👎  and I'll merge this PR as it is currently
